File: npalookup.pl
#!/usr/bin/perl

use strict;
use warnings;
use XML::Simple;
use LWP::Simple;
use Data::Dumper;

# Turn off output buffering
$|=1;

my $e164map = 0;

sub main
{
	if ( $ARGV[0] =~ /^$/ || $ARGV[1] =~ /^$/ ) {
		print "\nERROR: Missing parameters\n";
		print "usage: npalookup.pl [NPA] [NXX]\n";
		exit(1);
	}

	if ( $ARGV[2] =~ /^e164$/ ) {
		$e164map = 1;
	}

	print "Looking up NPA $ARGV[0] NXX $ARGV[1]\n";
	print "Downloading...\n";

	# Download local area codes and dial plan from localcallingguide.com
	my $localprefixes = get("http://localcallingguide.com/xmllocalprefix.php?npa=$ARGV[0]&nxx=$ARGV[1]");
	my $dialplan = get("http://localcallingguide.com/xmldialplan.php?npa=$ARGV[0]");

	my $parser = new XML::Simple;

	print ("Parsing...\n");
	my $dom = $parser->XMLin( $localprefixes );
	#print Dumper( $dom );
	my @entries = @{ $dom->{'lca-data'}->{'prefix'} };
	my %npas;

	print "\nLocal NPAs:\n";

	foreach my $npa ( @entries ) {
		push( @{ $npas{ $npa->{'npa'} } }, $npa->{'nxx'} );
		#$npas{ $npa->{'npa'} } = 1;
	}

	foreach my $npa ( keys %npas ) {
		print "  $npa\n";
	}

	print "\n";

	$dom = $parser->XMLin( $dialplan );

	# I keep finding odd spaces in these fields.
	foreach my $key ( keys $dom->{'dpdata'} ) {
		$dom->{'dpdata'}->{ $key } =~ s/\s//g;
	}

	#print Dumper( $dom );

	print "Dial Plan:\n";
	print "  Local (HNPA): " . $dom->{'dpdata'}->{'std_hnpa_local'} . "\n";
	print "  Local (FNPA): " . $dom->{'dpdata'}->{'std_fnpa_local'} . "\n";
	print "   Toll (HNPA): " . $dom->{'dpdata'}->{'std_hnpa_toll'} . "\n";
	print "   Toll (FNPA): " . $dom->{'dpdata'}->{'std_fnpa_toll'} . "\n";
	print "   Oper Assist: " . $dom->{'dpdata'}->{'std_oper_assis'} . "\n";
	print "\n";

	if ( $e164map ) {
		print "e164-maps:\n";
		foreach my $npa ( sort keys %npas ) {
			foreach my $nxx ( @{ $npas{ $npa } } ) {
				print " e164 \+1$npa$nxx....\n";
			}
		}
	}
}

main();

File: npalookup.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse
import logging
import pprint
import requests
import sys
import xmltodict

# Set up the logger
logger = logging.getLogger()
handler = logging.StreamHandler(stream=sys.stdout)
formatter_debug = logging.Formatter(
    '%(asctime)s [%(levelname)8s](%(funcName)s:%(lineno)d): %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S')
formatter = logging.Formatter(
    '%(asctime)s  %(message)s', datefmt='%Y-%m-%d %H:%M:%S')
handler.setFormatter(formatter)
logger.addHandler(handler)
logger.setLevel(logging.INFO)

# Used for large debug output
pp = pprint.PrettyPrinter(compact=True)


def init():
  parser = argparse.ArgumentParser(description="NPA Lookup")

  parser.add_argument("NPA", help="NPA to lookup")
  parser.add_argument("NXX", help="NXX to lookup")

  parser.add_argument("-d", "--debug", action="store_true",
                      help="Enable debug info")
  args = parser.parse_args()

  if args.debug:
    logger.setLevel(logging.DEBUG)
    handler.setFormatter(formatter_debug)
    logger.debug("Enabling DEBUG output")

  main(args)


def main(args):
  logger.debug("Arguments: {}".format(args))

  print(f"\nLooking up NPA: {args.NPA}  NXX: {args.NXX}\n")

  localprefixes = []

  lpdata = requests.get(f"https://localcallingguide.com/xmllocalprefix.php?npa={args.NPA}&nxx={args.NXX}")
  lpdict = xmltodict.parse(lpdata.content)

  error = lpdict['root']['lca-data'].get('error')
  if error:
    logger.debug(pp.pformat(lpdict))
    print(f"Error: {error}")

  prefixes = lpdict['root']['lca-data'].get('prefix')
  if prefixes:
    for prefix in prefixes:
      if prefix['npa'] not in localprefixes:
        localprefixes.append(prefix['npa'])

  if localprefixes:
    localprefixes.sort()
    # Remove NPA
    localprefixes.remove(args.NPA)
    # Move to beginning so we can copy into CUCDM directly
    localprefixes.insert(0, args.NPA)
    logger.debug("localprefixes is {}".format(pp.pformat(localprefixes)))
    print("Local NPAs: {}".format(",".join(localprefixes)))

  # Only check if we didn't get an error from above
  if not error:
    dpdata = requests.get(f"https://localcallingguide.com/xmldialplan.php?npa={args.NPA}")
    dpdict = xmltodict.parse(dpdata.content)
    dialplan = dpdict['root'].get('dpdata')
    if dialplan:
      print("\nDial Plan:")
      print("  Local (HNPA): {}".format(dialplan['std_hnpa_local']))
      print("  Local (FNPA): {}".format(dialplan['std_fnpa_local']))
      print("   Toll (HNPA): {}".format(dialplan['std_hnpa_toll']))
      print("   Toll (FNPA): {}".format(dialplan['std_fnpa_toll']))
      print("   Oper Assist: {}\n".format(dialplan['std_oper_assis']))


if __name__ == "__main__":
  init()
